src/rules-generator.ts

```diff
@@ -447,197 +451,186 @@ util.methods(Generator, {
+          throw new Error(errors.recursive + " (" + callee.methodName + ")");
         }
-        if (replaceExp()) {
-          exp.ref = ref;
-          exp.args = args;
-          // TODO: Get rid of this hack (for data.parent().val())
-          if (exp.ref.valueType === 'Snapshot') {
-            exp.valueType = 'Snapshot';
-          }
+        functionCalls[callee.methodName] = true;
+        let result = this.partialEval(fn.body, innerParams, functionCalls);
+        functionCalls[callee.methodName] = false;
+        return result;
+      }
+
+      // Can't expand function - but just expand the arguments.
+      if (!this.allowUndefinedFunctions) {
+        // TODO: Simplify type of accessor (not string | Exp).
+        var funcName = expCall.ref.type === 'ref'
+          ? <string> (<ast.ExpReference> expCall.ref).accessor
+          : (<ast.ExpVariable> expCall.ref).name;
+        if (!(funcName in this.symbols.schema['String'].methods ||
+              util.arrayIncludes(snapshotMethods, funcName))) {
+          this.fatal(errors.undefinedFunction + decodeExpression(expCall.ref));
         }
       }
 
-      return exp;
+      for (let i = 0; i < expCall.args.length; i++) {
+        expCall.args[i] = subExpression(expCall.args[i]);
+      }
+      // TODO: Get rid of this hack (for data.parent().val())
+      if (expCall.ref.valueType === 'Snapshot') {
+        expCall.valueType = 'Snapshot';
+      }
+
+      return expCall;
 
     // Expression types (like literals) than need no expansion.
     default:
       return exp;
     }
-  },
+  }
 
   // Builtin function - convert all 'this' to 'data' (from 'newData').
   // Args are function arguments, and params are the local (function) scope variables.
-  prior: function(args, params) {
+  prior(args: ast.Exp[], params: { [name: string]: ast.Exp; }): ast.Exp {
     var lastThisIs = this.thisIs;
     this.thisIs = 'data';
     var exp = this.partialEval(args[0], params);
     this.thisIs = lastThisIs;
     return exp;
-  },
+  }
 
   // Builtin function - current value of 'this'
-  getThis: function(args, params) {
+  getThis(args: ast.Exp[], params: { [name: string]: ast.Exp; }): ast.Exp {
     var result = ast.snapshotVariable(this.thisIs);
     return result;
-  },
+  }
 
   // Builtin function - convert string to RegExp
-  makeRegExp: function(args, params) {
+  makeRegExp(args: ast.Exp[], params: { [name: string]: ast.Exp; }) {
     if (args.length !== 1) {
       throw new Error(errors.application + "RegExp arguments.");
     }
-    var exp = this.partialEval(args[0], params);
+    var exp = <ast.ExpValue> this.partialEval(args[0], params);
     if (exp.type !== 'String' || !/\/.*\//.test(exp.value)) {
       throw new Error(errors.coercion + decodeExpression(exp) + " => RegExp");
     }
     return ast.regexp(exp.value);
-  },
+  }
 
   // Lookup globally defined function.
-  lookupFunction: function(ref) {
+  lookupFunction(ref: ast.ExpVariable | ast.ExpReference): {
```

Comment:
Out of scope for this PR - Bolt currently doesn't know about globally available functions or symbols, and just leaves them un-expanded.  But I agree would be better to have knowledge of that list locally to give error messages earlier (rather than get them on upload).